### PR TITLE
CIRCSTORE-331 Create actual-cost-record entity CRUD

### PR DESCRIPTION
**Pull request by Mykyta-Varenyk**
Resolves: https://issues.folio.org/browse/CIRCSTORE-331

### CODE REVIEW

File: descriptors/ModuleDescriptor-template.json [hunk 1]
```diff
@@ -440,6 +440,51 @@
...
+          "pathPattern":"/actual-cost-record-storage/actual-cost-records/{id}",
+          "permissionsRequired": [
+            "actual-cost-record-storage.actual-cost-records.item.get"
+          ]
+        },
+        {
+          "methods":[
+            "POST"
+          ],
+          "pathPattern":"/actual-cost-record-storage/actual-cost-records/{id}",
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply descriptors/ModuleDescriptor-template.json]: Removed

File: src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    Context vertxContext) {
+
+    PgUtil.post(ACTUAL_COST_RECORD_TABLE, entity, okapiHeaders, vertxContext,
+      PostActualCostRecordStorageActualCostRecordsResponse.class, asyncResultHandler);
+  }
+
+  public void getActualCostRecordStorageActualCostRecordsById(String id, String lang,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+    PgUtil.getById(ACTUAL_COST_RECORD_TABLE, ACTUAL_COST_RECORD_CLASS, id,
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java]: Identation added

File: src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import static org.folio.support.ModuleConstants.ACTUAL_COST_RECORD_CLASS;
+import static org.folio.support.ModuleConstants.ACTUAL_COST_RECORD_TABLE;
+
+public class ActualCostRecordAPI implements ActualCostRecordStorage {
+
+  public void getActualCostRecordStorageActualCostRecords(int offset, int limit, String query,
+    String lang, Map<String, String> okapiHeaders,
+    Handler<AsyncResult<Response>> asyncResultHandler, Context vertxContext) {
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java]: Fixed

File: src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    List<JsonObject> actualCostRecords = new ArrayList<>(
+      actualCostRecordClient.getMany("lossType==Aged to lost").getRecords());
+
+    assertThat(actualCostRecords, hasItems(hasSameProperties(createResult1),
+      hasSameProperties(createResult2)));
+
+    for (JsonObject current : actualCostRecords) {
+      actualCostRecordClient.deleteById(UUID.fromString(current.getString("id")));
+    }
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java:78]: Check added

File: src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    assertThat(createResult, hasSameProperties(actualCostRecord));
+
+    JsonObject updatedJson = createResult.put("lossType", DECLARED_LOST.value());
+
+    actualCostRecordClient.attemptPutById(updatedJson);
+
+    IndividualResource fetchedActualCostRecord = actualCostRecordClient.getById(updatedJson.getString("id"));
+
+    JsonObject fetchedJson = fetchedActualCostRecord.getJson();
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java]: Refactored

File: ramls/actual-cost-record.json [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      "description": "Fee fine type name",
+      "type": "string"
+    },
+    "metadata": {
+      "description": "Metadata about creation and changes, provided by the server (client should not provide)",
+      "type": "object",
+      "$ref": "raml-util/schemas/metadata.schema"
+    }
+  },
+  "additionalProperties": false
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record.json]: Required fields added

File: src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+      PostActualCostRecordStorageActualCostRecordsResponse.class, asyncResultHandler);
+  }
+
+  public void getActualCostRecordStorageActualCostRecordsById(String id, String lang,
+    Map<String, String> okapiHeaders, Handler<AsyncResult<Response>> asyncResultHandler,
+    Context vertxContext) {
+
+    PgUtil.getById(ACTUAL_COST_RECORD_TABLE, ACTUAL_COST_RECORD_CLASS, id,
+      okapiHeaders, vertxContext,
+      GetActualCostRecordStorageActualCostRecordsByIdResponse.class, asyncResultHandler);
```
Comment (roman-barannyk) [inline src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java:43]:
```suggestion
      okapiHeaders, vertxContext, GetActualCostRecordStorageActualCostRecordsByIdResponse.class, 
      asyncResultHandler);
```

File: src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import io.vertx.core.json.JsonObject;
+import lombok.SneakyThrows;
+import static org.folio.rest.jaxrs.model.ActualCostRecord.LossType.AGED_TO_LOST;
+import static org.folio.rest.jaxrs.model.ActualCostRecord.LossType.DECLARED_LOST;
+import static org.folio.rest.support.matchers.JsonMatchers.hasSameProperties;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.core.Is.*;
```
Comment (roman-barannyk) [inline src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java:35]: Please don't import all package

File: src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+  @SneakyThrows
+  public void canCreateAndGetAndUpdateActualCostRecord() {
+    JsonObject actualCostRecord = toJsonObject(createActualCostRecord());
+    JsonObject createResult = actualCostRecordClient.create(actualCostRecord).getJson();
+
+    assertThat(createResult, hasSameProperties(actualCostRecord));
+
+    JsonObject updatedJson = createResult.put("lossType", DECLARED_LOST.value());
+
+    actualCostRecordClient.attemptPutById(updatedJson);
```
Comment (roman-barannyk) [inline src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java:93]: why is it attempt? It seems like it should be successful put...

File: src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java (hunk 8 above)
Comment (Mykyta-Varenyk) [reply src/test/java/org/folio/rest/api/ActualCostRecordAPITest.java]: Fixed

File: src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java (hunk 7 above)
Comment (Mykyta-Varenyk) [reply src/main/java/org/folio/rest/impl/ActualCostRecordAPI.java]: Fixed

**Review by roman-barannyk — APPROVED** [review]

File: ramls/actual-cost-record.json [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+      "type": "string",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "userId": {
+      "description": "ID of the patron the item was lent to",
+      "type": "string",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "userBarcode": {
+      "description": "Barcode of the patron the item was lent to",
```
Comment (alexanderkurash) [inline ramls/actual-cost-record.json:17]: Barcode of the patron the actual cost record was created for

File: ramls/actual-cost-record.json [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+      "type": "string"
+    },
+    "loanId": {
+      "description": "Unique ID (generated UUID) of the loan",
+      "type": "string",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "itemLossType": {
+      "description": "How the item became lost",
```
Comment (alexanderkurash) [inline ramls/actual-cost-record.json:26]: Type of the item loss

File: ramls/actual-cost-record.json [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+      "type": "string",
+      "enum": [
+        "Aged to lost",
+        "Declared lost"
+      ]
+    },
+    "dateOfLoss": {
+      "description": "A date when the item was lost",
```
Comment (alexanderkurash) [inline ramls/actual-cost-record.json:34]: Date and time when the item was lost

File: ramls/actual-cost-record.json [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+          "readonly": true
+        }
+      }
+    },
+    "permanentItemLocation": {
+      "description": "Permanent item location is the default location, shelving location, or holding which is a physical place where items are stored, or an Online location.",
+      "type": "string"
+    },
+    "feeFineOwnerId": {
+      "description": "Fee fine owner ID",
```
Comment (alexanderkurash) [inline ramls/actual-cost-record.json:100]: Fee/fine

File: ramls/actual-cost-record.json [hunk 14]
```diff
@@ -0,0 +1,141 @@
...
+      "type": "string",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "feeFineOwner": {
+      "description": "Fee fine owner name",
```
Comment (alexanderkurash) [inline ramls/actual-cost-record.json:105]: Fee/fine

File: ramls/actual-cost-record.json [hunk 15]
```diff
@@ -0,0 +1,141 @@
...
+  "type": "object",
+  "description": "Actual cost record",
+  "properties": {
+    "id": {
+      "description": "Actual cost record ID",
+      "type": "string",
+      "$ref": "raml-util/schemas/uuid.schema"
+    },
+    "userId": {
+      "description": "ID of the patron the item was lent to",
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record.json]: Changed

File: ramls/actual-cost-record.json (hunk 13 above)
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record.json]: Changed

File: ramls/actual-cost-record.json (hunk 14 above)
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record.json]: Changed

File: ramls/actual-cost-record-storage.raml [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+        collection-item:
+          exampleItem: !include examples/actual-cost-record.json
+          schema: actual-cost-record
+      get:
+        description: "Looks up actual cost record"
+      put:
+        description: "Updates actual cost record"
+        is: [validate]
+      delete:
+        description: "Removes actual cost record"
```
Comment (alexanderkurash) [inline ramls/actual-cost-record-storage.raml:57]: Delete actual cost record by ID

File: ramls/actual-cost-record-storage.raml [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+      body:
+        application/json:
+          type: actual-cost-record
+    /{id}:
+      type:
+        collection-item:
+          exampleItem: !include examples/actual-cost-record.json
+          schema: actual-cost-record
+      get:
+        description: "Looks up actual cost record"
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record-storage.raml]: Changed

File: ramls/actual-cost-record-storage.raml [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+      put:
+        description: "Updates actual cost record"
```
[the remark this answers is not in the document]
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record-storage.raml]: Changed

**Review by alexanderkurash — APPROVED** [review]

File: ramls/actual-cost-record-storage.raml (hunk 16 above)
Comment (Mykyta-Varenyk) [reply ramls/actual-cost-record-storage.raml]: Changed

**Pull request merged**